File: KYDivisionPickerViewDemo_ObjC/KYDivisionPickerView/KYDivisionPickerView.m
//
//  KYDivisionPickerView.m
//  KYDivisionPickerViewDemo
//
//  Created by 徐开源 on 16/9/4.
//  Copyright © 2016年 kyxu. All rights reserved.
//

#import "KYDivisionPickerView.h"

@interface Location : NSObject
@property(nonatomic, assign) int num;
@property(nonatomic, strong) NSString* name;
@end
@implementation Location @end


@interface DivisionData : NSObject @end
@implementation DivisionData

+(NSDictionary*) divisionJson {
    
    NSString* path  = [[NSBundle mainBundle] pathForResource:@"KYDivisionPickerView_list" ofType:@"json"];
    // 将文件内容读取到字符串中，注意编码NSUTF8StringEncoding 防止乱码
    NSString* jsonString = [[NSString alloc] initWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
    // 将字符串写到缓冲区
    NSData* jsonData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    NSError* jsonError;
    NSDictionary* dic = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingMutableContainers error:&jsonError];
    return dic;
}

+(NSArray*) provinces {
    NSDictionary* dic = [DivisionData divisionJson];
    NSMutableArray* provincesArr = [[NSMutableArray alloc] init];
    // 添加数据
    NSEnumerator *enumerator = [dic keyEnumerator];
    NSString* key;
    while ((key = [enumerator nextObject])) {
        if ([key hasSuffix:@"0000"]) {
            Location* loc = [[Location alloc] init];
            loc.num = [key intValue];
            loc.name = [dic objectForKey: key];
            [provincesArr addObject: loc];
        }
    }
    [provincesArr sortUsingComparator:^NSComparisonResult(Location* obj1, Location* obj2) {
        return [obj1 num] > [obj2 num];
    }];
    return provincesArr;
}

+(NSMutableArray*) citiesWithProvinceNum:(int)provinceNum {
    NSDictionary* dic = [DivisionData divisionJson];
    // 确定索引
    NSString* provinceNumStr = [NSString stringWithFormat:@"%d", provinceNum];
    NSString* provinceNumPrefix = [provinceNumStr substringToIndex:2];
    NSMutableArray* citiesArr = [[NSMutableArray alloc] init];
    // 添加数据
    NSEnumerator *enumerator = [dic keyEnumerator];
    NSString* key;
    while ((key = [enumerator nextObject])) {
        if (([key hasPrefix:provinceNumPrefix]) && !([key hasSuffix:@"0000"]) && ([key hasSuffix:@"00"])) {
            Location* loc = [[Location alloc] init];
            loc.num = [key intValue];
            loc.name = [dic objectForKey: key];
            [citiesArr addObject: loc];
        }
    }
    if (citiesArr.count != 0) {
        [citiesArr sortUsingComparator:^NSComparisonResult(Location* obj1, Location* obj2) {
            return [obj1 num] > [obj2 num];
        }];
    }else {
        // 直辖市会获取不到下属的地级市，这时以直辖市本身作为地级市返回
        NSString* key = [provinceNumPrefix stringByAppendingString: @"0000"];
        Location* loc = [[Location alloc] init];
        loc.num = [key intValue];
        loc.name = [dic objectForKey: key];
        [citiesArr addObject: loc];
    }
    return citiesArr;
}

+(NSMutableArray*) countiesWithCityNum:(int)cityNum {
    NSDictionary* dic = [DivisionData divisionJson];
    // 确定索引
    NSString* cityNumStr = [NSString stringWithFormat:@"%d", cityNum];
    NSString* cityNumPrefix = [[NSString alloc] init];
    if ([cityNumStr hasSuffix: @"0000"]) { // 直辖市
        cityNumPrefix = [cityNumStr substringToIndex: 2];
    }else { // 地级市
        cityNumPrefix = [cityNumStr substringToIndex: 4];
    }
    NSMutableArray* countiesArr = [[NSMutableArray alloc] init];
    // 添加数据
    NSEnumerator *enumerator = [dic keyEnumerator];
    NSString* key;
    while ((key = [enumerator nextObject])) {
        if (([key hasPrefix:cityNumPrefix]) && !([key hasSuffix:@"00"])) {
            Location* loc = [[Location alloc] init];
            loc.num = [key intValue];
            loc.name = [dic objectForKey: key];
            [countiesArr addObject: loc];
        }
    }
    [countiesArr sortUsingComparator:^NSComparisonResult(Location* obj1, Location* obj2) {
        return [obj1 num] > [obj2 num];
    }];
    return countiesArr;
}

+(NSMutableArray*) streetsWithCountyNum:(int)countyNum {
    
    NSMutableArray* streetsArr = [[NSMutableArray alloc] init];
    
    // JSON
    NSString* jsonFileName = [NSString stringWithFormat:@"KYDivisionPickerView_%d", countyNum];
    NSString* path  = [[NSBundle mainBundle] pathForResource: jsonFileName ofType:@"json"];
    if (!path) {
        return streetsArr;
    }
    // 将文件内容读取到字符串中，注意编码NSUTF8StringEncoding 防止乱码
    NSString* jsonString = [[NSString alloc] initWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
    // 将字符串写到缓冲区
    NSData* jsonData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    NSError* jsonError;
    NSDictionary* dic = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingMutableContainers error:&jsonError];
    
    // 添加数据
    NSEnumerator *enumerator = [dic keyEnumerator];
    NSString* key;
    while ((key = [enumerator nextObject])) {
        Location* loc = [[Location alloc] init];
        loc.num = [key intValue];
        loc.name = [dic objectForKey: key];
        [streetsArr addObject: loc];
    }
    [streetsArr sortUsingComparator:^NSComparisonResult(Location* obj1, Location* obj2) {
        return [obj1 num] > [obj2 num];
    }];
    return streetsArr;
}

@end



@interface KYDivisionPickerView ()

@property(nonatomic, strong) NSArray* provinces;
@property(nonatomic, strong) NSMutableArray* cities;
@property(nonatomic, strong) NSMutableArray* counties;
@property(nonatomic, strong) NSMutableArray* streets;

@end

@implementation KYDivisionPickerView

#pragma mark - Life Cycle
-(instancetype)initWithFrame:(CGRect)frame {
    if ((self = [super initWithFrame:frame])) {
        
        _adjustsFontSizeToFitWidth = NO;
        _fontSize  = 14;
        _textColor = [UIColor blackColor];
        
        _provinces = [DivisionData provinces];
        _cities    = [DivisionData citiesWithProvinceNum: 110000]; // 默认传参北京市
        _counties  = [DivisionData countiesWithCityNum: 110000]; // 默认传参北京市
        _streets   = [DivisionData streetsWithCountyNum: 110101]; // 默认传参（北京市）东城区
        
        self.delegate = self;
        self.dataSource = self;
    }
    return self;
}

-(instancetype)initWithCoder:(NSCoder*)coder
{
    self = [super initWithCoder:coder];
    return self;
}


#pragma mark - UIPickerView
-(NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 4; // 省、市、区、街道
}

-(NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    switch (component) {
        case 0:
            return _provinces.count;
            break;
        case 1:
            return _cities.count;
            break;
        case 2:
            return _counties.count;
            break;
        case 3:
            return _streets.count;
            break;
        default:
            return 0;
            break;
    }
}

-(UIView *)pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row forComponent:(NSInteger)component reusingView:(UIView *)view {

    UILabel* label = [[UILabel alloc] initWithFrame: CGRectMake(8, 0, self.frame.size.width/4 - 16, 30)];
    label.adjustsFontSizeToFitWidth = _adjustsFontSizeToFitWidth;
    label.font = [UIFont systemFontOfSize: _fontSize];
    label.textAlignment = NSTextAlignmentCenter;
    Location* loc = [[Location alloc] init];
    
    switch (component) {
        case 0:
            if (row < _provinces.count) { loc = _provinces[row]; }
            break;
        case 1:
            if (row < _cities.count) { loc = _cities[row]; }
            break;
        case 2:
            if (row < _counties.count) { loc = _counties[row]; }
            break;
        case 3:
            if (row < _streets.count) { loc = _streets[row]; }
            break;
        default:
            break;
    }
    
    if (loc.name.length != 0) { label.text = loc.name; }
    return label;
}

-(void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    switch (component) {
        case 0:
            if (row < _provinces.count) {
                Location* loc = _provinces[row];
                [self updateComponentForCityWithProvinceNum:loc.num];
            }
            break;
        case 1:
            if (row < _cities.count) {
                Location* loc = _cities[row];
                [self updateComponentForCountyWithCityNum:loc.num];
            }
            break;
        case 2:
            if (row < _counties.count) {
                Location* loc = _counties[row];
                [self updateComponentForStreetWithCountyNum:loc.num];
            }
            break;
        case 3:
            [self passAddressStr];
            break;
        default:
            break;
    }
}


#pragma mark - Update
-(void)updateComponentForCityWithProvinceNum:(int)provinceNum { // 修改省，更新市、县、街道
    _cities = [DivisionData citiesWithProvinceNum: provinceNum];
    [self reloadComponent: 1];
    [self selectRow:0 inComponent:1 animated:YES];
    if (_cities.count != 0) {
        Location* loc = _cities[0];
        [self updateComponentForCountyWithCityNum: loc.num];
    }else {
        [_counties removeAllObjects];
        [_streets removeAllObjects];
        [self reloadComponent: 2];
        [self reloadComponent: 3];
        [self passAddressStr];
    }
}

-(void)updateComponentForCountyWithCityNum:(int)cityNum { // 修改市，更新县、街道
    _counties = [DivisionData countiesWithCityNum: cityNum];
    [self reloadComponent:2];
    [self selectRow:0 inComponent:2 animated:YES];
    if (_counties.count != 0) {
        Location* loc = _counties[0];
        [self updateComponentForStreetWithCountyNum:loc.num];
    }else {
        [_streets removeAllObjects];
        [self reloadComponent:3];
        [self passAddressStr];
    }
}

-(void)updateComponentForStreetWithCountyNum:(int)countyNum { // 修改县，更新街道
    _streets = [DivisionData streetsWithCountyNum:countyNum];
    [self reloadComponent:3];
    [self selectRow:0 inComponent:3 animated:YES];
    [self passAddressStr];
}

-(void)passAddressStr {
    NSString* provinceStr = [[NSString alloc] init];
    NSInteger row = [self selectedRowInComponent:0];
    if (row < 0) { row = 0;}
    if (row < _provinces.count) {
        Location* loc = _provinces[row];
        provinceStr = loc.name;
    }
    
    NSString* cityStr = [[NSString alloc] init];
    row = [self selectedRowInComponent:1];
    if (row < 0) { row = 0;}
    if (row < _cities.count) {
        Location* loc = _cities[row];
        cityStr = loc.name;
    }
    
    NSString* countyStr = [[NSString alloc] init];
    row = [self selectedRowInComponent:2];
    if (row < 0) { row = 0;}
    if (row < _counties.count) {
        Location* loc = _counties[row];
        countyStr = loc.name;
    }
    
    NSString* streetStr = [[NSString alloc] init];
    row = [self selectedRowInComponent:3];
    if (row < 0) { row = 0;}
    if (row < _streets.count) {
        Location* loc = _streets[row];
        streetStr = loc.name;
    }
    
    if (_divisionDelegate != nil) {
        // 代理方法传出 PickerView 对应的地址字符串
        [_divisionDelegate didGetAddressFromPickerViewWithProvinceName:provinceStr
                                                                  cityName:cityStr
                                                                countyName:countyStr
                                                                streetName:streetStr];
    }
}

@end

File: KYDivisionPickerViewDemo/KYDivisionPickerView/KYDivisionPickerView.swift
//
//  KYDivisionPickerView.swift
//  Administrative Division
//
//  Created by 徐开源 on 16/8/25.
//  Copyright © 2016年 kyxu. All rights reserved.
//

import UIKit

// 通过这个协议向 KYDivisionPickerView 的代理传递地址
protocol KYDivisionPickerViewDelegate {
    func didGetAddressFromPickerView(provinceName provinceName:String?, cityName:String?, countyName:String?, streetName:String?)
}


class KYDivisionPickerView: UIPickerView, UIPickerViewDelegate, UIPickerViewDataSource {
    
    // 可配置项
    var adjustsFontSizeToFitWidth = false
    var fontSize: CGFloat = 14
    var textColor: UIColor = UIColor.blackColor()
    
    // 代理传值
    var divisionDelegate: KYDivisionPickerViewDelegate? = nil
    
    // 数据来源
    private let provinces = DivisionData.provinces()
    private var cities    = DivisionData.cities(110000) // 默认传参北京市
    private var counties  = DivisionData.counties(110000) // 默认传参北京市
    private var streets   = DivisionData.streets(110101) // 默认传参（北京市）东城区
    
    
    // MARK: - Life Cycle
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.delegate = self
        self.dataSource = self
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    
    // MARK: - UIPickerView
    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int {
        return 4 // 省、市、区、街道
    }
    
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        switch component {
        case 0:
            return provinces.count
        case 1:
            return cities.count
        case 2:
            return counties.count
        case 3:
            return streets.count
        default:
            return 0
        }
    }
    
    func pickerView(pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusingView view: UIView?) -> UIView {
        let label = UILabel(frame: CGRectMake(8, 0, self.frame.width/4 - 16, 30))
        label.adjustsFontSizeToFitWidth = adjustsFontSizeToFitWidth
        label.font = UIFont.systemFontOfSize(fontSize)
        label.textColor = textColor
        label.textAlignment = .Center
        
        switch component {
        case 0:
            if row < provinces.count { label.text = provinces[row].name }
        case 1:
            if row < cities.count    { label.text = cities[row].name }
        case 2:
            if row < counties.count  { label.text = counties[row].name }
        case 3:
            if row < streets.count   { label.text = streets[row].name }
        default:
            break
        }
        return label
    }
    
    func pickerView(pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        switch component {
        case 0:
            updateComponentForCity(provinces[row].num)
        case 1:
            updateComponentForCounty(cities[row].num)
        case 2:
            updateComponentForStreet(counties[row].num)
        case 3:
            passAddressStr()
        default:
            break
        }
    }
    
    
    // MARK: - Update
    func updateComponentForCity(provinceNum:Int) { // 修改省，更新市、县、街道
        cities = DivisionData.cities(provinceNum)
        self.reloadComponent(1)
        self.selectRow(0, inComponent: 1, animated: true)
        if cities.count != 0 {
            updateComponentForCounty(cities[0].num)
        }else {
            counties.removeAll()
            streets.removeAll()
            self.reloadComponent(2)
            self.reloadComponent(3)
            passAddressStr()
        }
    }
    
    func updateComponentForCounty(cityNum:Int) { // 修改市，更新县、街道
        counties = DivisionData.counties(cityNum)
        self.reloadComponent(2)
        self.selectRow(0, inComponent: 2, animated: true)
        if counties.count != 0 {
            updateComponentForStreet(counties[0].num)
        }else {
            streets.removeAll()
            self.reloadComponent(3)
            passAddressStr()
        }
    }
    
    func updateComponentForStreet(countyNum:Int) { // 修改县，更新街道
        streets = DivisionData.streets(countyNum)
        self.reloadComponent(3)
        self.selectRow(0, inComponent: 3, animated: true)
        passAddressStr()
    }
    
    func passAddressStr() {
        var provinceStr: String? = nil
        var row = self.selectedRowInComponent(0)
        if row < 0 { row = 0 }
        if row < provinces.count {
            provinceStr = provinces[row].name
        }
        
        var cityStr: String? = nil
        row = self.selectedRowInComponent(1)
        if row < 0 { row = 0 }
        if row < cities.count {
            cityStr = cities[row].name
        }
        
        
        var countyStr: String? = nil
        row = self.selectedRowInComponent(2)
        if row < 0 { row = 0 }
        if row < counties.count {
            countyStr = counties[row].name
        }
        
        var streetStr: String? = nil
        row = self.selectedRowInComponent(3)
        if row < 0 { row = 0 }
        if row < streets.count {
            streetStr = streets[row].name
        }
        
        if let delegate = self.divisionDelegate {
            // 代理方法传出 PickerView 对应的地址字符串
            delegate.didGetAddressFromPickerView(
                provinceName: provinceStr, cityName: cityStr,
                countyName: countyStr, streetName: streetStr)
        }
    }
}


// 用这个类从 JSON 文件中获取内容
private class DivisionData {
    
    private struct Location {
        var num: Int
        var name: String
    }
    
    private class func divisionJson() -> JSON? {
        guard let divisionJsonFile = NSBundle.mainBundle().pathForResource("KYDivisionPickerView_list", ofType: "json")
            else {return nil}
        guard let data = NSData(contentsOfFile: divisionJsonFile)
            else {return nil}
        let json = JSON(data: data)
        return json
    }
    
    class func provinces() -> [Location] {
        guard let json = DivisionData.divisionJson() else {return [Location]()}
        var provincesArr = [Location]()
        // 添加数据
        for (index,subJson):(String, JSON) in json {
            if index.hasSuffix("0000") {
                if let num = Int(index) {
                    let loc = Location(num: num, name: String(subJson))
                    provincesArr.append(loc)
                }
            }
        }
        provincesArr.sortInPlace({$0.num < $1.num})
        return provincesArr
    }
    
    class func cities(provinceNum:Int) -> [Location] {
        guard let json = DivisionData.divisionJson() else {return [Location]()}
        // 确定索引
        let provinceNumPrefix = (String(provinceNum) as NSString).substringToIndex(2)
        var citiesArr = [Location]()
        // 添加数据
        for (index,subJson):(String, JSON) in json {
            if index.hasPrefix(provinceNumPrefix) && !index.hasSuffix("0000") && index.hasSuffix("00") {
                if let num = Int(index) {
                    let loc = Location(num: num, name: String(subJson))
                    citiesArr.append(loc)
                }
            }
        }
        if citiesArr.count != 0 {
            citiesArr.sortInPlace({$0.num < $1.num})
        }else {
            // 直辖市会获取不到下属的地级市，这时以直辖市本身作为地级市返回
            let index = provinceNumPrefix + "0000"
            if let num = Int(index) {
                let loc = Location(num: num, name: String(json[index]))
                citiesArr.append(loc)
            }
        }
        return citiesArr
    }
    
    class func counties(cityNum:Int) -> [Location] {
        guard let json = DivisionData.divisionJson() else {return [Location]()}
        // 确定索引
        let cityNumStr = String(cityNum)
        let cityNumPrefix: String
        if cityNumStr.hasSuffix("0000") { // 直辖市
            cityNumPrefix = (cityNumStr as NSString).substringToIndex(2)
        }else { // 地级市
            cityNumPrefix = (cityNumStr as NSString).substringToIndex(4)
        }
        var countiesArr = [Location]()
        // 添加数据
        for (index,subJson):(String, JSON) in json {
            if index.hasPrefix(cityNumPrefix) && !index.hasSuffix("0") {
                if let num = Int(index) {
                    let loc = Location(num: num, name: String(subJson))
                    countiesArr.append(loc)
                }
            }
        }
        countiesArr.sortInPlace({$0.num < $1.num})
        return countiesArr
    }
    
    class func streets(countyNum:Int) -> [Location] {
        guard let divisionJsonFile = NSBundle.mainBundle().pathForResource("KYDivisionPickerView_\(countyNum)", ofType: "json")
            else {return [Location]()}
        guard let data = NSData(contentsOfFile: divisionJsonFile)
            else {return [Location]()}
        let json = JSON(data: data)
        var streetsArr = [Location]()
        // 添加数据
        for (index,subJson):(String, JSON) in json {
            if let num = Int(index) {
                let loc = Location(num: num, name: String(subJson))
                streetsArr.append(loc)
            }
        }
        streetsArr.sortInPlace({$0.num < $1.num})
        return streetsArr
    }

}